BUG: algorithms.factorize moves null values when sort=False

rhshadrach opened the pull request:
- [x] closes #46584 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

In the example below, the result_index has nan moved even though sort=False. This is the order that will be in any groupby reduction result and the reason why transform currently returns wrong results.

```
df = pd.DataFrame({'a': [1, 3, np.nan, 1, 2], 'b': [3, 4, 5, 6, 7]})
print(df.groupby('a', sort=False, dropna=False).grouper.result_index)

# main
Float64Index([1.0, 3.0, 2.0, nan], dtype='float64', name='a')
# this PR
Float64Index([1.0, 3.0, nan, 2.0], dtype='float64', name='a')
```

cc @jbrockmendel @jreback

rhshadrach committed:
BUG: algos.factorizes moves null values when sort=False

rhshadrach commented on pandas/core/algorithms.py:
```diff
@@ -728,25 +733,31 @@ def factorize(
 
     if not isinstance(values.dtype, np.dtype):
         # i.e. ExtensionDtype
+        assert dropna or sort
         codes, uniques = values.factorize(na_sentinel=na_sentinel)
```
Currently not hit by any tests, but I think would be if e.g Int64 was used in the grouper. This will need fixing.

rhshadrach commented on pandas/tests/test_algos.py:
```diff
@@ -436,13 +436,13 @@ def test_factorize_na_sentinel(self, sort, na_sentinel, data, uniques):
-                np.array([0, 2, 1, 0], dtype=np.dtype("intp")),
-                np.array(["a", "b", np.nan], dtype=object),
+                np.array([0, 1, 2, 0], dtype=np.dtype("intp")),
+                np.array(["a", None, "b"], dtype=object),
```
The change from np.nan to None here was an unintended side-effect; is this undesirable?

mroeschke replied:
IMO I think this could be called a bug fix since now a user can preserve None in their input data when using `pd.factorize`. Probably good to add a whatsnew note for this chage.

rhshadrach replied:
Thanks! I completely forgot about this open question. Will do.

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -728,25 +733,31 @@ def factorize(
 
     if not isinstance(values.dtype, np.dtype):
         # i.e. ExtensionDtype
+        assert dropna or sort
```
can you add a comment about why this assertion should hold

rhshadrach replied:
Ah, this was just to demonstrate no tests hit here; will be adding tests that hit here with both dropna and sort being False. See OP for details.

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -1044,7 +1044,7 @@ def factorize(self, na_sentinel: int = -1) -> tuple[np.ndarray, ExtensionArray]:
 
         codes, uniques = factorize_array(
-            arr, na_sentinel=na_sentinel, na_value=na_value
+            arr, na_sentinel=na_sentinel, na_value=na_value, ignore_na=True
```
is ignore_na just always going to be True?

rhshadrach replied:
I think no - this part (EAs) is still to come.

rhshadrach commented:
@jbrockmendel - I've implemented the EA paths. There is a bit of a discrepancy between pd.factorize and EA's factorize (as well as factorize_array). Namely, the former can take `na_sentinel=None` whereas the latter two have `na_sentinel` being integers. Because of this, I added the `dropna` to the EA factorize as well as factorize_array. Could also call this argument `ignore_na`, but I think that isn't used by the public API so `dropna` seemed better.

I wonder if we should add `dropna` to pd.factorize and deprecate `na_sentinel` being None there.

rhshadrach committed:
Implementation for pyarrow < 4.0

rhshadrach committed:
fixup

rhshadrach committed:
fixups

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach committed:
test fixup

rhshadrach committed:
type-hints

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach commented:
@jbrockmendel - gentle ping.

jreback reviewed (approved):
over to you @jbrockmendel

jbrockmendel commented on pandas/core/algorithms.py:
```diff
@@ -541,7 +547,11 @@ def factorize_array(
-        values, na_sentinel=na_sentinel, na_value=na_value, mask=mask
+        values,
+        na_sentinel=na_sentinel,
+        na_value=na_value,
```
for the list of things to improve signature-clarity-wise, na_sentinel vs na_value is confusing

rhshadrach replied:
Is it just the names or descriptions in the docstring? If it's the name, perhaps na_code instead of na_sentinel. Or are you thinking more we shouldn't have both as arguments?

As I was working on this, I was wondering what use case na_sentinel has. In other words, what's the downside of having na_sentile always be -1 and hence not an argument.

jbrockmendel commented on pandas/core/arrays/masked.py:
```diff
@@ -869,16 +869,43 @@ def searchsorted(
         assert uniques.dtype == self.dtype.numpy_dtype, (uniques.dtype, self.dtype)
 
-        uniques_ea = type(self)(uniques, np.zeros(len(uniques), dtype=bool))
+        # Make room for a null value if we're not ignoring it and it exists
```
would it make sense to share any of this with the ArrowArray version?  not for this PR, but could have a TODO

rhshadrach replied:
Yes, will add. Once we drop support for pyarrow < 4.0 we won't need this logic in ArrowArray, but 4.0 is only a year old at this point so that will be a while.

jbrockmendel commented on pandas/core/arrays/string_.py:
```diff
@@ -382,6 +382,13 @@ def _values_for_factorize(self):
+    @classmethod
+    def _from_factorized(cls, values, original):
+        assert values.dtype == original._ndarray.dtype
+        # When dropna (i.e. ignore_na) is False, can get -1 from nulls
```
is there any way we could avoid this?

rhshadrach replied:
Thanks - yes. Changed _values_for_factorize from -1 to None.

jbrockmendel commented on pandas/core/groupby/grouper.py:
```diff
@@ -657,9 +657,10 @@ def group_index(self) -> Index:
             # we make a CategoricalIndex out of the cat grouper
-            # preserving the categories / ordered attributes
+            # preserving the categories / ordered attributes;
+            # doesn't (yet) handle dropna=False
```
GH ref for the "yet"?

rhshadrach replied:
Opened #46909, will add a reference in this comment.

jbrockmendel commented:
> I wonder if we should add dropna to pd.factorize and deprecate na_sentinel being None there. Can also go the other route here and allow na_sentinel being None in EA's factorize; no strong opinion but I find na_sentinel=None meaning "Don't drop null values from uniques; and code nulls as -1" not very clear.

I agree on the lack of clarity, would like to see this cleaned up (separate PR(s))

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach committed:
improvements

rhshadrach commented:
Thanks for the comments @jbrockmendel, ready for another go.

> I agree on the lack of clarity, would like to see this cleaned up (separate PR(s))

Opened #46910

rhshadrach committed:
remove override of _from_factorized in string_.py

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach committed:
Rework na_sentinel/dropna/ignore_na

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach committed:
fixups

rhshadrach committed:
fixup for pyarrow < 4.0

rhshadrach committed:
Merge branch 'main' into factorize_na

rhshadrach committed:
whatsnew note

rhshadrach commented:
@jbrockmendel - Based on the discussion in #46910, I reworked the handling of na_sentinel in factorize. Instead of adding `dropna`, this PR now allows `na_sentinel=None`.

rhshadrach committed:
doc fixup

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach commented:
@jbrockmendel - gentle ping

jbrockmendel commented:
@rhshadrach thanks for being patient with me.

I'm not wild about adding na_sentiel=None to EA.factorize (would rather deprecate na_sentiel xref #46910), but if you say it is necessary to fix a bug (at least until such a deprecation is enforced), I'll take your word for it

rhshadrach commented:
@jbrockmendel Something is necessary, but I like the idea of deprecating specifying na_sentinel in pd.factorize and changing EA's factorize to match what pd.factorize will look like in 2.0. Let's wait on discussion in #46910. If we do go this route, I'll throw up a PR for the deprecation and we can return to this PR after.

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst
	pandas/core/arrays/arrow/array.py
	pandas/core/arrays/base.py
	pandas/core/arrays/datetimelike.py
	pandas/core/arrays/masked.py
	pandas/core/arrays/sparse/array.py

rhshadrach committed:
fixups

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

rhshadrach commented:
@jbrockmendel - friendly ping

jbrockmendel commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -278,6 +278,8 @@ Other enhancements
 - :meth:`DatetimeIndex.astype` now supports casting timezone-naive indexes to ``datetime64[s]``, ``datetime64[ms]``, and ``datetime64[us]``, and timezone-aware indexes to the corresponding ``datetime64[unit, tzname]`` dtypes (:issue:`47579`)
 - :class:`Series` reducers (e.g. ``min``, ``max``, ``sum``, ``mean``) will now successfully operate when the dtype is numeric and ``numeric_only=True`` is provided; previously this would raise a ``NotImplementedError`` (:issue:`47500`)
 - :meth:`RangeIndex.union` now can return a :class:`RangeIndex` instead of a :class:`Int64Index` if the resulting values are equally spaced (:issue:`47557`, :issue:`43885`)
+- The method :meth:`.ExtensionArray.factorize` can now be passed ``use_na_sentinel=False`` for determining how null values are to be treated. (:issue:`46601`)
```
"be passed" -> "takes" or "accepts", i.e. avoid passive voice

rhshadrach replied:
Thanks - fixed.

rhshadrach committed:
fixup whatsnew note

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

jbrockmendel commented:
> friendly ping

Thanks for your patience.  This is near the top of my queue.

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst
	pandas/tests/groupby/test_groupby_dropna.py

rhshadrach commented on pandas/core/algorithms.py:
```diff
@@ -753,38 +758,58 @@ def factorize(
             # GH#46910 - TimelikeOps has deprecated signature
             codes, uniques = values.factorize(  # type: ignore[call-arg]
-                use_na_sentinel=True
+                use_na_sentinel=na_sentinel is not None
```
Yes - but how much depends on what you mean by "this". There is a lot of playing around with arguments (what I think you're highlighting here) that will all go away. But what I regard as the main complication this PR introduces, noted in the TODO immediately below, will not go away just with the deprecation.

For that TODO, some changes to `safe_sort` are necessary. For the bulk of cases the changes are straightforward and only require very small changes. However if you have multiple Python types in an object array (e.g. float and string) with np.nan and None, I haven't yet to find a good way to sort. I plan to revisit this in the next few days.

rhshadrach replied:
I looked at safe_sort again, and @phofl already solve much of the issue in #47331. I opened a PR into this branch here to see what the execution of the TODO mentioned above would look like: https://github.com/rhshadrach/pandas/pull/2

However, changing safe_sort in this way will induce another behavior change:

```
df = pd.DataFrame({'a': ['x', None, np.nan], 'b': [1, 2, 3]})
print(df.groupby('a', dropna=False).sum())

# main
     b
a     
x    1
NaN  5

# feature
      b
a      
x     1
None  2
NaN   3
```

No other tests besides the one changed failed locally. While I do think this is a bugfix, I'd like to study more what impact it has on concat/reshaping/indexing and I think it may need some discussion. In particular, I don't think it should be done in this PR.

rhshadrach replied:
@jbrockmendel - friendly ping.

jbrockmendel replied:
> While I do think this is a bugfix, I'd like to study more what impact it has on concat/reshaping/indexing and I think it may need some discussion. In particular, I don't think it should be done in this PR.

Agreed both that the "feature" behavior looks more correct and that it should be done separate from this PR.

Taking another look now.

mroeschke replied:
(Coming in fairly cold), I think I'm getting lost too here. I am not fully comprehending why we check `            na_sentinel == -1 or na_sentinel is None` and then use `use_na_sentinel=na_sentinel is not None`

rhshadrach replied:
Yea - there are some gymnastics here no doubt. The idea behind this block is to avoid using catch_warnings since it's possible.

Old API: na_sentinel is either an integer or None
New API: use_na_sentinel is False or True

The correspondence is:

use_na_sentinel False is equivalent to na_sentinel being None
use_na_sentinel True is equivalent to na_sentinel is -1

Note there is no option in the new API for na_sentinel being anything other than -1 or None in the old API. So we can use the new argument precisely when (a) the function has said argument and (b) na_sentinel is either -1 or None. In such a case, the correspondence from na_sentinel to use_na_sentinel is given by `use_na_sentinel = na_sentinel is not None`.

mroeschke replied:
Perfect, just the explanation I needed. Thanks!

rhshadrach replied:
I'll add this correspondence as a comment to the top of this function; we can remove it when the deprecation is enforced.

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

jbrockmendel commented:
@mroeschke im flailing on reviewing this.  can you tag in?

mroeschke reviewed (commented):
One whatsnew comment, one comprehension comment, merge conflict; otherwise, LGTM

rhshadrach committed:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

rhshadrach committed:
whatsnew note; comment on old vs new API

mroeschke reviewed (approved):
LGTM just one merge conflict

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into factorize_na

 Conflicts:
	doc/source/whatsnew/v1.5.0.rst

mroeschke commented:
Great! Thanks for the effort on this @rhshadrach
